src/allocator.rs

```diff
@@ -332,13 +342,31 @@ impl Allocator {
 
     pub fn new_pair(&mut self, first: NodePtr, rest: NodePtr) -> Result<NodePtr, EvalErr> {
         let idx = self.pair_vec.len();
-        if idx == MAX_NUM_PAIRS {
+        if idx >= MAX_NUM_PAIRS - self.num_ghost_pairs {
             return err(self.nil(), "too many pairs");
         }
         self.pair_vec.push(IntPair { first, rest });
         Ok(NodePtr::new(ObjectType::Pair, idx))
     }
 
+    // this code is used when we are simulating pairs with a vec locally
+    // in the deserialize_br code
+    // we must maintain parity with the old deserialize_br code so need to track the skipped pairs
+    pub fn add_ghost_pair(&mut self, amount: usize) -> Result<(), EvalErr> {
+        if MAX_NUM_PAIRS - self.num_ghost_pairs - self.pair_vec.len() < amount {
+            return err(self.nil(), "too many pairs");
+        }
+        self.num_ghost_pairs += amount;
+        Ok(())
+    }
+
+    // this code is used when we actually create the pairs that were previously skipped ghost pairs
+    pub fn remove_ghost_pair(&mut self, amount: usize) -> Result<(), EvalErr> {
+        // currently let this panic with overflow if we go below 0 to debug if/where it happens
+        self.num_ghost_pairs -= amount;
```

Comment:
it might make sense to have a `debug_assert!(self.num_ghost_pairs >= amount)`.
But I suppose that's implied in the integer wrapping

Reply:
Added